proposals/defer-hydration.md

```diff
@@ -0,0 +1,90 @@
+# Defer Hydration Protocol
+
+An open protocol for controlling hydration on the client.
+
+Author: Justin Fagnani
+
+Status: Draft
+
+Last update: 2021-06-24
+
+# Background
+
+In server-side rendered (SSR) applications, the process of a component running code to re-associate its template with the server-rendered DOM is called "hydration". The defer-hydration protocol is design to allow controler over hydration to solve two related problems:
+
+1. Interoperable incremental hydration across web components. Componentents should not automatically hydrate upon being defined, but wait for a signal from their parent or other coordinator.
+2. Hydration ordering independent of definition order. Because components usually depend on data from their parent, and the parent won't usually set data on a child until it's hydrated, we need hydration to occur in a top-down order.
```

Comment:
I guess this is maybe a partial answer for (1)? Should these be separate points, if so?

Reply:
The more I think about it, I don't think I agree with this point (2), cc @Westbrook.  I would expect that in most cases an element's initial data has been serialized as attributes and doesn't need the parent to pass that information. The exception would be data that isn't serialized, like complex objects, but I don't think that's really the norm.

Reply:
The https://github.com/w3c/webcomponents-cg has a breakout session to discuss Community Protocols, that could certainly include this, at the beginning of next month. Would you be interested in joining the call? Would be great to get your thoughts on this, and many other of the topics we're discussing right now "in person"!

Reply:
Yeah I could probably attend that.

Just for clarity, I'm not against this proposal, just trying to understand the scenarios in which it is needed. The text could help by clarifying that a bit more.

My thinking is that a typical usage would not need this because:

1. Static imports take care of making sure children are hydrated first.
2. Even if hydration happens out-of-order state should be serialized in attributes.